src/Lorentz/Contracts/BaseDAO/Types.hs

```diff
@@ -3,90 +3,332 @@
+import Util.Markdown
+
+------------------------------------------------------------------------
+-- Configuration
+------------------------------------------------------------------------
+
+data Config proposalMetadata = Config
+  { cDaoName :: Text
+  , cDaoDescription :: Markdown
+  , cUnfrozenTokenMetadata :: FA2.TokenMetadata
+  , cFrozenTokenMetadata :: FA2.TokenMetadata
+
+  , cProposalCheck :: forall s
+      . (ProposeParams proposalMetadata) & s
+    :-> Bool & s
+  -- TODO: Explain ^
+
+  , cRejectedProposalReturnValue :: forall s
+      . (Proposal proposalMetadata) & s
+    :-> ("slash_amount" :! Natural) & s
+  -- TODO: Explain ^
+
+  , cDecisionLambda :: forall s
+      . (Proposal proposalMetadata) & (Storage proposalMetadata) & s
+    :-> (List Operation, Storage proposalMetadata) & s
+  -- ^ The decision lambda is executed based on a successful proposal.
+  }
+
+defaultConfig :: Config pm
+defaultConfig = Config
+  { cDaoName = "BaseDAO"
+  , cDaoDescription = [md|"BaseDAO description"|]
+  , cUnfrozenTokenMetadata = FA2.TokenMetadata
+      { FA2.tmTokenId = unfrozenTokenId
+      , FA2.tmSymbol = [mt|unfrozen_token|]
+      , FA2.tmName = [mt|Unfrozen Token|]
+      , FA2.tmDecimals = 8
+      , FA2.tmExtras = mempty
+      }
+  , cFrozenTokenMetadata = FA2.TokenMetadata
+      { FA2.tmTokenId = frozenTokenId
+      , FA2.tmSymbol = [mt|frozen_token|]
+      , FA2.tmName = [mt|Frozen Token|]
+      , FA2.tmDecimals = 8
+      , FA2.tmExtras = mempty
+      }
+  , cProposalCheck = do
+      drop; push True
+  , cRejectedProposalReturnValue = do
+      drop; push (0 :: Natural); toNamed #slash_amount
+  , cDecisionLambda = do
+      drop; nil; pair
+  }
+------------------------------------------------------------------------
+-- Operators
+------------------------------------------------------------------------
 
 type Operators = BigMap (Address, (Address, FA2.TokenId)) ()
 
-type Ledger = BigMap (Address, FA2.TokenId) Natural
+------------------------------------------------------------------------
+-- Ledger
+------------------------------------------------------------------------
+
+type Ledger = BigMap LedgerKey LedgerValue
+
+type LedgerKey = (Address, FA2.TokenId)
+
+type LedgerValue = Natural
+
+------------------------------------------------------------------------
+-- Storage/Parameter
+------------------------------------------------------------------------
 
 -- | Storage of the FA2 contract
-data Storage = Storage
+data Storage (proposalMetadata :: Kind.Type) = Storage
   { sLedger       :: Ledger
   , sOperators    :: Operators
   , sTokenAddress :: Address
   , sAdmin        :: Address
   , sPendingOwner :: Maybe Address
+
+  , sProposalIdCounter :: Natural
+  , sVotingPeriod :: VotingPeriod
+  , sQuorumThreshold :: QuorumThreshold
+
+  , sProposals :: BigMap ProposalId (Proposal proposalMetadata)
+  , sProposalsIdSortByDate :: [ProposalId] -- Newest first (probably change to set)
```

Comment:
Having unbounded lists in the storage is dangerous. For example, if there are too many proposals, the contract may become unusable because storage deserialization will become too expensive. So we should explicitly limits lengths of all strict data types. For example, we can require that there is no more than 500 active proposals and it's impossible for one proposal to have more than 1000 votes. I think it can be part of compile-time configuration (I mean these numbers: 500 and 1000). Then we should mention it in the specification.

If there are other unbounded types, we should limit them as well. For example, `QuorumThreshold` is `Natural` and `nat` type is unbounded, so it can be arbitrarily large as well. Even though `sQuorumThreshold` is available only to the admin who we trust, I think we should limit it as a safety measure.

Reply:
I have added the bounded value in `Config` as below: 

```haskell
data Config = Config pm 
  { -- Bounded
  , cMaxProposals :: Natural
  , cMaxVotes :: Natural -- including both upvotes and downvotes
  , cMaxQuorumThreshold :: Natural -- Must be smaller than or equal to MaxVotes or else the threshold could be never met.
  , cMinQuorumThreshold :: Natural
  , cMaxVotingPeriod :: Natural -- Maximum of seconds allow to be set
  , cMinVotingPeriod :: Natural
  }

defaultConfig :: Config pm
defaultConfig =
  { 
  , cMaxVotingPeriod = 60 * 60 * 24 * 30 -- 1 month
  , cMinVotingPeriod = 60 * 60 -- 1 hour

  , cMaxQuorumThreshold = 1000
  , cMinQuorumThreshold = 1

  , cMaxVotes = 1000

  , cMaxProposals = 500
  }
```

I also added a check using the abouve bounded-value in `propose`, `vote`, `setQuorumThreshold` and `setVotingPeriod` entrypoints as well. Let me know if this is good enough or not, Ivan.

Reply:
There new checks look ok. But there is still some work that needs to be done in this PR. Please ping me when it's ready.